tests/lib/external/snapd-testing-tools/utils/log-parser

```diff
@@ -235,97 +238,126 @@ class Operation:
         self.task = task
         self.source_line = source_line
 
-    def __repr__(self):
+    def __repr__(self) -> str:
         return self.source_line
 
-    def __dict__(self):
+    def _some_dict(self) -> dict[str, typing.Any]:
         return {
-            'type': self.type,
-            'date': self.date,
-            'time': self.time,
-            'verb': self.verb,
-            'task': self.task,
-            'extra': self.extra
-            }
+            "type": self.type,
+            "date": self.date,
+            "time": self.time,
+            "verb": self.verb,
+            "task": self.task,
+            "extra": self.extra,
+        }
 
 
 class LogReader:
     """
     LogReader manages the spread log, it allows to read, export and print
     """
-    def __init__(self, filepath, output_type, lines_limit, store_setup, error_rules, debug_rules):
+
+    def __init__(
+            self, filepath: str, output_type: str, lines_limit: int, store_setup: str, error_rules: list[str], debug_rules: list[str]
+    ):
         self.filepath = filepath
         self.output_type = output_type
         self.lines_limit = lines_limit
         self.store_setup = store_setup
-        self.lines = []
+        self.lines = list[str]()
         self.iter = 0
-        self.full_log = []
+        self.full_log = list[Result|Info|Operation|Detail|Action|str]()
         self.error_rules = [Rule(rule) for rule in error_rules]
         self.debug_rules = [Rule(rule) for rule in debug_rules]
 
-    def __repr__(self):
-        return str(self.__dict__())
+    def __repr__(self) -> str:
+        return str(self._some_dict())
 
-    def __dict__(self):
-        return {'full_log': self.full_log}
+    def _some_dict(self) -> dict[str, typing.Any]:
+        return {"full_log": self.full_log}
 
-    def print_log(self, details, results):
+    def print_log(self, details: str, results: str) -> None:
         if not self.full_log:
             return
 
         # Print the details
         if details == ALL:
-            print(''.join(str(x) for x in self.full_log))
+            print("".join(str(x) for x in self.full_log))
         elif details == NONE:
             pass
         elif details == ERROR:
-            print(''.join(str(x) for x in self.full_log if x.type == INFO and
-                  x.info_type == ERROR_TYPE))
+            print(
+                "".join(
+                    str(x)
+                    for x in self.full_log
+                    if isinstance(x, Info) and x.info_type == helper.ERROR
```

Comment:
This is also something I've remarked in my patch, using `x.type` is weird and unwarranted: we have `isinstance` for that.